File: libremarkable/refresh.c
#include "lib.h"
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <sys/mman.h>

#define min(a,b) (((a)<(b))?(a):(b))
#define max(a,b) (((a)>(b))?(a):(b))

// 0 is an invalid update_marker value
unsigned gen = 1;

// rect=NULL for full-screen refresh
uint32_t remarkable_framebuffer_refresh(remarkable_framebuffer* fb,
                                        update_mode refresh_mode,
                                        waveform_mode waveform,
                                        display_temp temp,
                                        mxcfb_dithering_mode dither_mode,
                                        int flags,
                                        unsigned int quant_bit,
                                        mxcfb_alt_buffer_data* alt_buffer_data,
                                        unsigned y, unsigned x,
                                        unsigned height, unsigned width) {
  if (fb == NULL)
    return -1;

  // TODO: So the drawing is twice as densely packed as refreshing horizontally?
  x = x / 2;

  mxcfb_update_data data = {0};
  data.update_region.top = max(min(y, fb->vinfo.yres - 1), 0);
  data.update_region.left = max(min(x, fb->vinfo.xres - 1), 0);

  if (data.update_region.left + width >= fb->vinfo.xres)
    data.update_region.width = fb->vinfo.xres - data.update_region.left;
  else 
    data.update_region.width = width;

  if (data.update_region.top + height >= fb->vinfo.yres)
    data.update_region.height = fb->vinfo.yres - data.update_region.top;
  else 
    data.update_region.height = height;

  data.waveform_mode = waveform;
  data.temp = temp;

  data.update_mode = refresh_mode;
  data.update_marker = gen++;
  data.dither_mode = 0; // entirely unused since epdc_v1 is used, not v2.
  data.quant_bit = quant_bit;
  data.flags = flags;
  data.dither_mode = dither_mode;

  if (alt_buffer_data != NULL)
    data.alt_buffer_data = *alt_buffer_data;

  int res = ioctl(fb->fd, REMARKABLE_PREFIX(MXCFB_SEND_UPDATE), &data);
  if (res != 0) {
    printf("ioctl(.., MXCFB_SEND_UPDATE) = %d\n", res);
    return -1;
  }
  
  // Return the marker so that the caller can wait for it to finish drawing if needed
  return data.update_marker;
}

uint32_t remarkable_framebuffer_wait_refresh_marker(remarkable_framebuffer* fb, uint32_t marker) {
  if (fb == NULL)
    return -1;

  // TODO: Collusion test (2nd value) is an output param here. It's value might be useful.
	mxcfb_update_marker_data mdata = { marker, 0 };
	ioctl(fb->fd, REMARKABLE_PREFIX(MXCFB_WAIT_FOR_UPDATE_COMPLETE), &mdata);
	return mdata.collision_test;
}

File: librustpad/src/refresh.rs
use libc;

use std::os::unix::io::AsRawFd;
use std::sync::atomic::Ordering;

use fb;
use mxc_types;
use mxc_types::{mxcfb_update_marker_data, mxcfb_update_data, mxcfb_rect};

impl<'a> fb::Framebuffer<'a> {
	pub fn refresh_region(
        &mut self,
        update_region: mxc_types::mxcfb_rect,
        update_mode: mxc_types::update_mode,
        waveform_mode: mxc_types::waveform_mode,
        temperature: mxc_types::display_temp,
        dither_mode: mxc_types::dither_mode,
        quant_bit: i32,
        flags: u32
    ) -> u32 {
    	let whole = mxcfb_update_data {
            update_mode: update_mode as u32,
            update_marker: *self.marker.get_mut() as u32,
            waveform_mode: waveform_mode as u32,
            temp: temperature as i32,
            flags: flags,
            quant_bit: quant_bit,
            dither_mode: dither_mode as i32,
            update_region: update_region,
            ..Default::default()
        };
        let pt: *const mxcfb_update_data = &whole;
        unsafe {
            libc::ioctl(self.device.as_raw_fd(), mxc_types::MXCFB_SEND_UPDATE, pt);
        }
        // TODO: Do proper compare and swap
        self.marker.swap(whole.update_marker + 1, Ordering::Relaxed);
        return whole.update_marker;
    }
	
    pub fn refresh(
        &mut self,
        y: usize,
        x: usize,
        height: usize,
        width: usize,
        update_mode: mxc_types::update_mode,
        waveform_mode: mxc_types::waveform_mode,
        temperature: mxc_types::display_temp,
        dither_mode: mxc_types::dither_mode,
        quant_bit: i32,
        flags: u32,
    ) -> u32 {
    	return self.refresh_region(mxcfb_rect {
                top: y as u32,
                left: x as u32,
                height: height as u32,
                width: width as u32,
            }, 
    		update_mode,
    		waveform_mode,
    		temperature,
    		dither_mode,
    		quant_bit,
    		flags);
    }

    pub fn wait_refresh_complete(&mut self, marker: u32) {
        let mut markerdata = mxcfb_update_marker_data {
            update_marker: marker,
            collision_test: 0,
        };
        unsafe {
            libc::ioctl(
                self.device.as_raw_fd(),
                mxc_types::MXCFB_WAIT_FOR_UPDATE_COMPLETE,
                &mut markerdata,
            );
        };
        // TODO: Return collision test -- kernel updates it to the next marker's collision data
    }
}
